grammar.js

```diff
@@ -1,337 +1,411 @@
+/**
+ * @file FIRRTL grammar for tree-sitter
+ * @author Aliaksei Chapyzhenka <alex.drom@gmail.com>
+ * @author Amaan Qureshi <amaanq12@gmail.com>
+ * @author Andrew Young <youngar17@gmail.com>
+ * @license Apache-2.0
+ * @see {@link https://www.chisel-lang.org/firrtl|official website}
+ * @see {@link https://github.com/chipsalliance/firrtl|official repository}
+ * @see {@link https://github.com/chipsalliance/firrtl-spec|official spec}
+ */
+
+/* eslint-disable arrow-parens */
+/* eslint-disable camelcase */
+/* eslint-disable-next-line spaced-comment */
+/// <reference types="tree-sitter-cli/dsl" />
+// @ts-check
+
 'use strict';
 
-function optseq() {
-  return optional(prec.left(seq.apply(null, arguments)));
+/**
+  * Creates a rule to match one or more of the rules separated by the separator
+  *
+  * @param {string} sep - The separator to use.
+  * @param {Rule} rule
+  *
+  * @return {SeqRule}
+  *
+  */
+function sep1(sep, rule) {
+  return seq(rule, repeat(seq(sep, rule)));
 }
 
-function repseq() {
-  return repeat(prec.left(seq.apply(null, arguments)));
-}
+module.exports = grammar({
+  name: 'firrtl',
 
-function sep1(separator, rule) {
-  return prec.left(seq(
-    rule,
-    repeat(prec.left(seq(separator, rule)))
-  ));
-}
+  externals: $ => [
+    $._newline,
+    $._indent,
+    $._dedent,
+  ],
 
-const rules = {
+  extras: $ => [
+    $.comment,
+    /\s/,
+  ],
 
-  source_file: $ => repeat($.circuit),
+  inline: $ => [
+    $.keyword_identifier,
+  ],
 
-  circuit: $ => seq(
-    'circuit', $.id, ':', optional($.info),
-    optseq($._indent, repeat($.module))
-  ),
+  supertypes: $ => [
+    $.expression,
+    $.statement,
+  ],
 
-  module: $ => choice(
-    seq(
-      'module', $.id, ':', optional($.info),
-      optseq(
-        $._indent,
-        repeat($.port),
-        repeat($.simple_stmt),
-        $._dedent
-      )
-    ),
-    seq(
-      'extmodule', $.id, ':', optional($.info),
-      optseq(
-        $._indent,
-        repeat($.port),
-        optional($.defname),
-        repeat($.parameter),
-        $._dedent
-      )
-    )
-  ),
-
-  port: $ => seq(
-    $.dir, $.id, ':', $.type, optional($.info), $._newline
-  ),
-
-  dir: $ => choice('input', 'output'),
-
-  type: $ => choice(
-    seq(
-      choice('UInt', 'SInt', 'Analog'),
-      optseq('<', $.intLit, '>')
+  word: $ => $.identifier,
+
+  rules: {
+    source_file: $ => repeat($.circuit),
+
+    circuit: $ => seq(
+      'circuit', $.identifier, ':', optional($.info),
+      optional(seq($._indent, repeat($.module), $._dedent)),
     ),
-    seq(
-      'Fixed',
-      optseq('<', $.intLit, '>'),
-      optseq('<', '<', $.intLit, '>', '>')
+
+    module: $ => choice(
```

Comment:
@amaanq 
Uh, This is total revamping of content, style, organization, and structure.
Could you summarize factual changes?
What bugs it fixes?
What features where added?
For example, does it fix #2 ?

Reply:
Been a while now, but let me summarize it from memory as best as I can

1. mux and other reserved keywords could not be used as identifiers, however in the main repository (firrtl/regress/FPU.fir) mux is used as an identifier, so adding the keyword_identifier rule aliased to identifier and inlining it solves this.

2. better distinguishing of escape sequences

3. add optional const qualifier support in port

4. extract any rules in statements into their own rule, which then enables setting statement as a supertype. supertypes allow us to use them in a query, but hides them in a node - inline doesn't allow us to use them in a query as well.

5. yes, when-else is fixed, see:
![image](https://user-images.githubusercontent.com/29718261/224205886-b02f76f1-4974-4033-8486-29b089e52ebb.png)

6. make queries (imo) perfect to neovim treesitter standards (we will standardize this with helix/other editors soon, as Max the tree-sitter creator & others have agreed last-wins is best - aka Neovim convention)

7. use snake_case over camelcase as is convention in tree-sitter grammars

8. ensure every FIRRTL file parses successfully and correctly

9. fill out docs as best as possible in manifests and libraries/packages

10. add swift package

11. format other files like rust, swift header, c/c++ based on appropriate standards

12. update tests & ci as unix-like work together just fine, then windows is its own thing

Reply:
I also added support to vim/neovim for auto detection of FIRRTL files so anyone who installs the grammar (after I add it to nvim-treesitter after this is merged) will have highlights, code folding, and indents working out of the box :) 

See:
![Peek 2023-03-09 21-23](https://user-images.githubusercontent.com/29718261/224207230-f676b03e-9a83-4a4b-9e25-06c9e4d77c36.gif)